### PR TITLE
fix: Dispose `TextBoxComponent` image cache properly

### PR DESCRIPTION
**Pull request by spydon**
# Description

Since we didn't dispose of the `Image` object that was used as a cache in `TextBoxComponent` we had a memory leak.

Side note: I wonder if it is really worth using `PictureRecorder` in that class.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

Fixes: #1462 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/test/components/text_box_component_test.dart [hunk 1]
```diff
@@ -37,5 +37,39 @@ void main() {
           ),
       );
     });
+
+    flameGame.test(
+      'internal image is disposed when component is removed',
+      (game) async {
```
Comment (erickzanardo) [inline packages/flame/test/components/text_box_component_test.dart:43]:
```suggestion
      (game) {
```

**Review by erickzanardo — APPROVED** [review]
One nit, but LGTM already

File: packages/flame/test/components/text_box_component_test.dart [hunk 2]
```diff
@@ -37,5 +37,39 @@ void main() {
...
+        game.render(canvas);
+        game.remove(c);
+        game.update(0);
+        expect(imageCache != null, isTrue);
+        expect(imageCache!.debugDisposed, isTrue);
+        expect(c.cache, null);
+      },
+    );
+
+    flameGame.test(
```
Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:59]:
```suggestion
    testWithFlameGame(
```

Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:59]:
```suggestion
    testWithFlameGame(
```

File: packages/flame/test/components/text_box_component_test.dart [hunk 3]
```diff
@@ -37,5 +37,39 @@ void main() {
...
+        final c = TextBoxComponent(text: 'foo bar');
+
+        await game.ensureAdd(c);
+        final imageCache = c.cache;
+
+        final canvas = MockCanvas();
+        game.render(canvas);
+        game.remove(c);
+        game.update(0);
+        expect(imageCache != null, isTrue);
```
Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:53]:
```suggestion
        expect(imageCache, isNotNull);
```

File: packages/flame/test/components/text_box_component_test.dart [hunk 4]
```diff
@@ -37,5 +37,39 @@ void main() {
...
+    flameGame.test(
+      'internal image is redrawn when component is re-added',
+      (game) async {
+        final c = TextBoxComponent(text: 'foo bar');
+
+        await game.ensureAdd(c);
+        game.remove(c);
+        game.update(0);
+        await game.ensureAdd(c);
+        await c.mounted;
```
Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:68]:
```suggestion
        expect(c.isMounted, true);
```

File: packages/flame/test/components/text_box_component_test.dart [hunk 5]
```diff
@@ -37,5 +37,39 @@ void main() {
...
+
+        expect(c.cache != null, isTrue);
```
Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:70]:
```suggestion
        await null;
        expect(c.cache, isNotNull);
```

File: packages/flame/test/components/text_box_component_test.dart [hunk 6]
```diff
@@ -37,5 +37,39 @@ void main() {
           ),
       );
     });
+
+    flameGame.test(
```
Comment (st-pasha) [inline packages/flame/test/components/text_box_component_test.dart:41]:
```suggestion
    testWithFlameGame(
```

**Comment by st-pasha** [comment]
I think there's still a possibility of a leak here if `redraw()` is called multiple times in a row. Not sure how to test for this though.

I believe the following version of `redraw` would be more robust:
```dart
  Future<void> redraw() async {
    final newSize = _recomputeSize();
    final image = await _fullRenderAsImage(newSize);
    cache?.dispose();
    cache = image;
    size = newSize;
  }
```

File: packages/flame/test/components/text_box_component_test.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/test/components/text_box_component_test.dart]: Haha I shouldn't have committed this right before I went to bed, it's needed due to the `await`.

**Pull request merged**